addons/project/models/project_task.py

```diff
@@ -1273,40 +1216,49 @@ def write(self, vals):
 
         # sends an email to the 'Task Creation' subtype subscribers
         # When project_id is changed
+        project_link_per_task_id = {}
         if vals.get('project_id'):
             project = self.env['project.project'].browse(vals.get('project_id'))
             notification_subtype_id = self.env['ir.model.data']._xmlid_to_res_id('project.mt_project_task_new')
             partner_ids = project.message_follower_ids.filtered(lambda follower: notification_subtype_id in follower.subtype_ids.ids).partner_id.ids
             if partner_ids:
-                link_per_project_id = {}
-                for task in self:
-                    if task.project_id:
-                        project_link = link_per_project_id.get(task.project_id.id)
-                        if not project_link:
-                            project_link = link_per_project_id[task.project_id.id] = task.project_id._get_html_link(title=task.project_id.display_name)
-                        project_link_per_task_id[task.id] = project_link
+                link_per_project_id = {
+                    project.id: project._get_html_link(title=project.display_name)
+                    for project in self.project_id
+                }
+                project_link_per_task_id = {
+                    task.id: link_per_project_id[task.project_id.id]
+                    for task in self
+                    if task.project_id
+                }
+        if vals.get('parent_id') is False:
+            additional_vals['display_in_project'] = True
+
+        # write changes
+        if self.env.su or not self.env.user._is_portal():
+            vals.update(additional_vals)
+        elif additional_vals:
+            super(ProjectTask, self.sudo()).write(additional_vals)
         result = super().write(vals)
-        if portal_can_write:
-            super(ProjectTask, self_no_sudo).write(vals_no_sudo)
 
         if 'user_ids' in vals:
             self._populate_missing_personal_stages()
 
         # user_ids change: update date_assign
         if 'user_ids' in vals:
-            for task in self:
+            for task in self.sudo():
                 if not task.user_ids and task.date_assign:
                     task.date_assign = False
                 elif 'date_assign' not in vals and task.id in task_ids_without_user_set:
                     task.date_assign = now
 
         # rating on stage
         if 'stage_id' in vals and vals.get('stage_id'):
-            self.filtered(lambda x: x.project_id.rating_active and x.project_id.rating_status == 'stage')._send_task_rating_mail(force_send=True)
+            self.sudo().filtered(lambda x: x.project_id.rating_active and x.project_id.rating_status == 'stage')._send_task_rating_mail(force_send=True)
```

Comment:
TBH, I don't really understand why we no longer add the sudo when it is really needed instead of adding it even if it is not needed. 😕

Reply:
same as above https://github.com/odoo/odoo/pull/203891#discussion_r2032704987